File: problems/leetcode/0200/212/212.py
# 212. Word Search II
# https://leetcode.com/problems/word-search-ii/description/

from typing import *

# 1. words로 Trie 생성
# 2. 칸마다 돌아다니면서 인접한 칸을 확인해서 Trie에 있으면 계속 나아가기
# O(m * n * 3^k)

class TrieNode:
    def __init__(self):
        self.value = None
        self.children = dict()

    def insert(self, s: str) -> None:
        node = self

        for ch in s:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]

        node.value = s

class Solution:
    def __init__(self):
        self.root = TrieNode()
        self.results = set()
        self.deltas = [(1, 0), (-1, 0), (0, 1), (0, -1)]

    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        for word in words:
            self.root.insert(word)

        m, n = len(board), len(board[0])
        for i in range(m):
            for j in range(n):
                visited = set()
                self.traverse(board, m, n, i, j, self.root, visited)

        return list(self.results)

    def traverse(self, board: List[List[str]], m: int, n: int, r: int, c: int, node: TrieNode, visited: set[tuple[int, int]]) -> None:
        if board[r][c] not in node.children.keys():
            return

        nextNode = node.children[board[r][c]]
        if nextNode.value:
            self.results.add(nextNode.value)

        visited.add((r, c))
        for dr, dc in self.deltas:
            nextR, nextC = r + dr, c + dc

            if 0 <= nextR < m and 0 <= nextC < n and (nextR, nextC) not in visited:
                self.traverse(board, m, n, nextR, nextC, nextNode, visited)

        visited.remove((r, c))

File: problems/leetcode/0200/212/212.cpp
// 212. Word Search II
// https://leetcode.com/problems/word-search-ii/description/

using namespace std;

#include <string>
#include <unordered_set>
#include <vector>

class TrieNode {
public:
  bool end = false;
  TrieNode *nodes['z' - 'a' + 1];

  TrieNode() {
    for (int i = 0; i < 'z' - 'a' + 1; i++) {
      nodes[i] = nullptr;
    }
  }
};

class Solution {
public:
  const int dr[4] = {0, 0, 1, -1};
  const int dc[4] = {1, -1, 0, 0};

  TrieNode *root = new TrieNode();
  int m = 0, n = 0;
  unordered_set<string> found;
  vector<vector<bool>> visited;

  bool isValidCoordinate(int r, int c) {
    return 0 <= r && r < m && 0 <= c && c < n;
  }

  void insert(string &s) {
    TrieNode *node = root;

    for (char ch : s) {
      if (!node->nodes[ch - 'a']) {
        node->nodes[ch - 'a'] = new TrieNode();
      }
      node = node->nodes[ch - 'a'];
    }

    node->end = true;
  }

  void dfs(vector<vector<char>> &board, int r, int c, TrieNode *node,
           string &current) {
    if (!node->nodes[board[r][c] - 'a']) {
      return;
    }

    current.push_back(board[r][c]);
    visited[r][c] = true;

    TrieNode *next = node->nodes[board[r][c] - 'a'];
    if (next->end) {
      found.insert(current);
    }

    for (int i = 0; i < 4; i++) {
      int newR = r + dr[i], newC = c + dc[i];
      if (isValidCoordinate(newR, newC) && !visited[newR][newC]) {
        dfs(board, newR, newC, next, current);
      }
    }

    current.pop_back();
    visited[r][c] = false;
  }

  vector<string> findWords(vector<vector<char>> &board, vector<string> &words) {
    m = board.size();
    n = board[0].size();
    visited = vector<vector<bool>>(m, vector<bool>(n, false));

    for (string &word : words) {
      insert(word);
    }

    string current;
    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        dfs(board, i, j, root, current);
      }
    }

    return vector<string>(found.begin(), found.end());
  }
};
